### ISSUE TITLE
ParameterReassignment false positive for C++ reference out parameters

### ISSUE THREAD

**Issue by cbiffle** [oclint/oclint#315]
The ParameterReassignment check fires for this idiomatic C++ code:

```
void swap(int & a, int & b) {
  int t = a;
  a = b;
  b = t;
}
```

I would argue that (perhaps under the control of a flag) assigning to a _reference_ parameter should not result in a warning.

**Comment by UnrealQuester** [comment]
I don't have a strong opinion on this one, but I rarely have out parameters that get reassigned.
If you have a few function you want to disable the warning on, you can [disable the warning](http://docs.oclint.org/en/stable/howto/suppress.html#annotation-syntax) at the method level.

**Comment by cbiffle** [comment]
I'm not sure what you mean by "reassigned" -- assigning a reference out-parameter *once* triggers the warning.  Surely you assign your out-parameters once. :-)

**Comment by UnrealQuester** [comment]
> I'm not sure what you mean by "reassigned" -- assigning a reference out-parameter once triggers the warning. Surely you assign your out-parameters once. :-)

Yes, that's what I meant.

**Comment by cbiffle** [comment]
The current "reassignment" warning is firing on a single assignment to a reference parameter, as shown in the example in my report above.  This is probably correct for a non-reference parameter, but I argue that it doesn't make sense in its current state.

*Two* assignments to a single reference parameter?  That'd be suspicious.

**Comment by UnrealQuester** [comment]
We can change the Rule to only apply to non-reference parameters, but I would like to hear some more opinions on this. What do you think, @ryuichis?

**Comment by cbiffle** [comment]
Well, again, it would be useful to detect *reassignment* of reference parameters.  (The rule's name is arguably wrong; it's actually detecting any assignment.)

But I leave the semantics up to y'all.

**Comment by AbigailBuccaneer** [comment]
I think *re*assignment of reference parameters is suspicious only if the first one is unnecessary: that should be found by a dead store analyser like Clang static analyzer's `deadcode.DeadStores` checker.

**Comment by ryuichi-assistant** [comment]
There hasn't been any activity on this issue recently. Due to the high number of incoming GitHub notifications, we have to clean some of the old issues, as many of them have already been resolved with the latest updates.

Please make sure to update to the latest version and check if that solves the issue. Let us know if that works for you by adding a comment.

**Comment by ryuichi-assistant** [comment]
This issue will be auto-closed because there hasn't been any activity for a few months. Feel free to open a new one if you still experience this problem.
